### PR TITLE
[SPARK-10562][SQL] support mixed case partitionBy column names for tables stored in metastore

### PR DESCRIPTION
**Pull request by cloud-fan**
This PR implements approach 1 of https://github.com/apache/spark/pull/9226#issuecomment-150272837

### CODE REVIEW

**Comment by yhuai** [comment]
@liancheng Can you also take a look?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1410,4 +1410,12 @@ class SQLQuerySuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
       }
     }
   }
+
+  test("SPARK-10562: partition by column with mixed case name") {
+    withTable("tbl10562") {
+      val df = Seq(2012 -> "a").toDF("Year", "val")
+      df.write.partitionBy("Year").saveAsTable("tbl10562")
+      checkAnswer(sql("SELECT year FROM tbl10562"), Row(2012))
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Can we add a test to use `yEAr`?

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Also, let's also have a test using the partition column in the predicate.

**Comment by liancheng** [comment]
Please see my comments in #9226.

**Comment by cloud-fan** [comment]
closing in favor of https://github.com/apache/spark/pull/9226

**Pull request closed without merging**